src/clojure_tic_tac_toe/minimax.clj

```diff
@@ -23,40 +24,51 @@
     (apply (max-or-min-based-on-player player) scores)))
 
 
-(defn- create-map-of-moves-and-scores
-  [moves scores]
-  (zipmap moves scores))
-
-(defn- get-pairs-with-target-score
-  [moves-and-scores target-score]
-  (filter #(-> % val (= target-score)) moves-and-scores))
-
-(defn- get-move-from-pairs
-  [pairs]
-  (first (first pairs)))
-
-(defn- get-move-based-on-score
-  [moves scores target-score]
-  ((comp get-move-from-pairs
-         #(get-pairs-with-target-score % target-score)
-         create-map-of-moves-and-scores)
-   moves scores))
-
-(defn minimax-move-and-score
-  [game-state {:keys [winning-moves valid-moves] :as initial-data}]
-  (if (:finished? game-state)
-    (let [score (game_handler/calculate-score game-state winning-moves)]
-      {:move :invalid-move, :score score})
-    (let [moves (game_handler/get-available-moves game-state valid-moves)
-          new-game-states (create-new-game-states game-state initial-data moves)
-          moves-and-scores (map #(minimax-move-and-score % initial-data)
-                                new-game-states)
-          scores (get-scores moves-and-scores)
-          score (get-score-based-on-player scores game-state)
-          move (get-move-based-on-score moves scores score)]
-      {:move move, :score score})))
+(defn- create-score-info-list
+  [scores moves total-moves]
+  (map #(assoc {} :score %1 :move %2 :total-moves %3)
+       scores moves total-moves))
+
+
+(defn- get-maps-with-target-score
+  [score-info-list score]
+  (filter #(= score (:score %)) score-info-list))
+
+(defn- sort-maps-by-total-moves
+  [maps]
+  (sort-by :total-moves (vec maps)))
+
+(defn- extract-move-from-first-map
+  [sorted-maps]
+  (:move (first sorted-maps)))
+
+(defn get-optimal-move
```

Comment:
should this be a private function?

Reply:
I would like it to be private, since it doesn't need to be a part of the public API. However, I also wanted to keep the tests, to clearly show what the function does. Ideally, I would extract get-optimal-move into a separate namespace. However, I am fine with keeping the function in minimax.clj for now (if I end up adding more functions to minimax, then at that point I'll clean up the namespace further).

Reply:
Are you not able to test this through the public facing function that uses this? If you can, let's try and do that instead

Reply:
I made the function private and removed the tests for that function.